### DISCUSSION TITLE
Zweiten/ Dritten MPPT hinzufügen

### DISCUSSION THREAD

**Discussion by Snusme83** [hoylabs/OpenDTU-OnBattery#797] [Q&A] [answered]
Servus @ all, super Projekt schon mal vorab.

Mit der neuen Version (2024.03.23) soll es ja möglich sein, einen zweiten Victron Laderegler hinzuzufügen.
Muss ich quasi genau die selbe Kabelverbindung zur OpenDTU-OnBattery nochmal herstellen und der zweite Laderegler wird mir angezeigt oder bedarf es da noch anderer Veränderungen?

Allerdings weiß ich dann nicht, an welche Pins ich den zweiten Laderegler anschließen muss! Ich denke das Pin-Mapping passt bei mir nicht, oder?

Danke schonmal für eure Antworten.

![Screenshot 2024-03-24 082239](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/cf96ae87-5044-4a61-8baf-adfa0f74472d)
![Screenshot 2024-03-24 082326](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/af9938aa-d86f-4f6d-8cd6-ef5039563363)

**Comment by SW-Niko** [comment, marked as answer]
Hallo @Snusme83,
du benötigst noch rx2. Siehe hier. 
https://github.com/helgeerbe/OpenDTU-OnBattery/pull/706#issuecomment-2000224310

**Reply by Snusme83** [reply]
Die sind bei mir aber nicht versehen mit einem Pin!
![Screenshot 2024-03-24 134558](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/81714dba-4131-47e5-bd3e-8b3962bbb6fd)

**Reply by spcqike** [reply]
Dann musst du dann musst du die pin mapping anpassen

https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/Device-Profiles

**Reply by SW-Niko** [reply]
Richtig. Und darum musst du erst bei deinem ESP32 einen freien Pin suchen und konfigurieren.
Wie das geht findest du in der Wiki. 
PS: Momentan gehen maximal 2 MPPTs. (Wenn 2 Serielle Schnittstellen frei sind)

**Reply by Snusme83** [reply]
Danke Jungs, nur nochmal zum Verständnis:
Ich sichere meine pin_mapping.json, bearbeite mit dem Editor und füge rx 2 und tx2 hinzu mit Hinterlegung der Pinnummer, speichere, stelle diese gespeicherte pin_mapping.json wieder her und wenn ich den 2. Smartsolar angeschlossen hab, müsste dieser erscheinen in der OpenDTU-onBattery?

**Reply by SW-Niko** [reply]
Ja genau. Den tx2 benötigst du nicht und wenn du ihn trotzdem konfigurierst dann keinen "GPI only" verwenden.

**Reply by Snusme83** [reply]
Passt das quasi so?

![Screenshot 2024-03-24 150138](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/30d2ee44-9e10-406d-a5d7-4d6feaca5b41)

**Reply by Snusme83** [reply]
Ok, 18 und 19 war keine gute Idee. Jetzt glaub ich hab ich es verstanden.

Ist GND überhaupt nötig für den Smartsolar?

**Reply by schlimmchen** [reply]
In der Tat, 18 und 19 sind ja schon in Gebrauch.

> Ist GND überhaupt nötig für den Smartsolar?

Ja. Du darfst andernfalls keine zuverlässige Kommunikation erwarten. Und eigentlich sollst du einen ADUM1201 dazwischenklemmen.

**Reply by Snusme83** [reply]
Nach langem probieren habe ich es jetzt hinbekommen. Hab am Anfang die Pins 35 und 34 verwendet.  34 und 35 sind "GPI only"!
Diese sollten laut SW-Niko nicht verwendet werden. Jedes Mal, wenn ich die pin_mapping.json wiederhergestellt habe, hing sich der ESP32 auf und nix ging mehr.

Zudem habe ich bei meiner pin_mapping.json im huawei auf "-1" gesetzt, da da einer der beiden (32, 33) vergeben war.

Was jetzt ursächlich dafür war, dass sich der ESP aufhängt, kann ich nicht sagen. Gibt es eine Möglichkeit den ESP wieder zum Leben zu erwecken, ohne neu flashen zu müssen?
![Screenshot 2024-03-25 bearbeitet](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/7a8eabb3-0906-4e0c-a25a-eb0b128bb772)
![Screenshot 2024-03-25 bearbeitet2](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/b4b59cf0-0f77-4adc-9774-6654fa33d253)


@schlimmchen: Wird es irgendwann noch eine Erweiterung auf 3 unterstützte Laderegler geben?

**Reply by schlimmchen** [reply]
Nicht ohne Weiteres. 3 Hardware-UARTs hat nur der ESP32-S3, und auch die werden zur Zeit noch nicht genutzt von OpenDTU-OnBattery. Hier ist eine Abstraktion nötig (die ich plane einzubauen). Damit der Anschluss von 3 VE.Direct MPPTs überhaupt möglich wird auf anderen ESP32 Modellen, muss eine fuktionierende SoftwareSerial Klasse her. Es gibt zwar eine passende Bibliothek, aber die braucht entweder unzumutbar viel Speicher, oder wir müssen noch je einen Prozess drumherumbasteln, der asynchron zur loop() die Daten aus dem SoftwareSerial herausnimmt.

Also leider nein, damit kannst du sobald nicht rechnen. Am ehesten noch wenn du einen ESP32-S3 einsetzt (OpenDTU FUSION Board beispielsweise) und die o.g. Abstraktion abwartest.

Natürlich ist dann mit 3 MPPTs alles an Hardware-UARTs ausgereizt und du kannst kein JK BMS oder Victron Smart Shunt oder SDM Power Meter mehr betreiben.

**Reply by Snusme83** [reply]
Danke für die Rückmeldung. JK BMS, Victron Smart Shunt oder SDM Power Meter nutze ich nicht oder ist ein Volkszähler ein SDM Power Meter? Im Grunde funktioniert unser "Balkonkraftwerk" mit Speicher (6 x Module 370 Wp, 3 x Victron Smartsolar, 2 x Pylontech US3000C) dank der OpenDTU-onBattery genauso, wie ich es von Anfang an realisieren wollte - und zwar vollautomatisch.
Was mir halt in der Anzeige fehlt ist der 3. Laderegler bzw. dessen Werte/ Erträge.

Was natürlich noch eine Möglichkeit wäre, von 3 Victron Smartsolar 100/20 auf 2 Victron Smartsolar 150/30 umzubauen.

Dann wäre die Anlage absolut perfekt. Vielleicht mache ich das nochmal!🤷🏻‍♂️

**Reply by rs3hc** [reply]
Wenn du wirklich 3 brauchst, würde ich einen Victron Cerbo GX hinzufügen.
Der verwaltet dann die Victron Geräte komplett und kann die Daten an OpenDTU weitergeben.

@schlimmchen wurde das schon gemacht, bzw wird das unterstützt?

**Reply by Snusme83** [reply]
Ich hatte vor dem OpenDTUon Battery einen Raspberry Pi 4 als Cerbo GX Ersatz um mir in Venus OS alles zu visualisieren (SoC der Speicher, Ertrag 3 x Laderegler bzw. alle 3 Gesamt, Volkszähler,...)
![IMG_9610](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/40e7542f-d30e-4081-821a-772ba19c7d0a)

Leider funktionierten aber mit einer bestimmten opendtuon Battery die Anzeigen nicht mehr alle!
Da mir die openDTU-on Battery nahezu alles anzeigt und ich jetzt die Anlage vollautomatisch laufen lassen kann, habe ich Venus OS aufgegeben.

**Reply by Snusme83** [reply]
So ist es für mich absolut in Ordnung, aber der dritte Laderegler fehlt halt!
![IMG_0140](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/21980aa9-587d-4575-b75b-968f0ebb7f7a)

**Reply by Snusme83** [reply]
![IMG_0142](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/164622660/c9e05ba2-8544-43ec-81e6-eafde9a2c367)

**Comment by Manos1966** [comment]
Hallo @Snusme83 

> Hab am Anfang die Pins 35 und 34 verwendet. 34 und 35 sind "GPI only"!

rein Informativ, anbei die GPIO Ports
Dunkel "Orange" sind die "Digital Input Ports" (GPIO36,39,34,35)
Als "Input" hättest du sie, so fern ich weiss (...und das ist nicht viel), verwenden koennen, also ein 
            "rx2": 34, (oder 35,36,39)
            "tx2": -1
sollte funktionieren. 
Ich habe gerade meine MPPT-Emulation benutzt (es ist Nacht, daher der echte Victron is aus) auf "rx2": 34. Die OpenDTU konnte die Daten lesen.

![ESP32 IO](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/129660360/80b6bff3-3204-4b82-8ecc-6494b1981ded)

**Reply by Snusme83** [reply]
Danke für die Rückmeldung. Habe jetzt für meine beiden Laderegler die gelben/ hell Orangen GPIO's für tx, rx, tx2 und rx2 verwendet, da ich das Display eh nicht mehr verwende.

Klappt so einwandfrei!🤓
